storehaus-cassandra/src/main/scala/com/twitter/storehaus/cassandra/cql/AbstractCQLCassandraCompositeStore.scala

```diff
@@ -0,0 +1,294 @@
+import com.datastax.driver.core.querybuilder.{BuiltStatement, Clause, QueryBuilder, Update, Select}
+import com.twitter.storehaus.{IterableStore, QueryableStore, ReadableStore, Store}
+import com.websudos.phantom.CassandraPrimitive
+import java.util.concurrent.{Executors, TimeUnit}
+import scala.annotation.tailrec
+import scala.collection.JavaConversions._
+import scala.collection.mutable.ArrayBuffer
+import scala.language.implicitConversions
+import shapeless._
+import HList._
+import ops.hlist.Mapper
+import ops.hlist.Mapped
+import poly._
+import Nat._
+import UnaryTCConstraint._
+import org.slf4j.LoggerFactory
+
+object AbstractCQLCassandraCompositeStore {
+  
+  /**
+   * used for mapping the keys in the HLists to Strings
+   */
+  object keyStringMapping extends (CassandraPrimitive ~>> String) {
+    override def apply[T](c: CassandraPrimitive[T]) = c.cassandraType
+  }
+  
+  /**
+   * used to map over an example key to implicit serializers
+   */
+  object cassandraSerializerCreation extends Poly1 {
+    implicit def default[T : CassandraPrimitive] = at[T](_ => implicitly[CassandraPrimitive[T]])
+  }
+
+  /**
+   * creates an HList of serializers (CasasandraPrimitives) given an example 
+   * key HList and pulls serializers out of implicit scope.
+   */
+  def getSerializerHListByExample[KL <: HList, SL <: HList](keys: KL)(implicit mapper: Mapper.Aux[cassandraSerializerCreation.type, KL, SL]) : SL = {
+    keys.map(cassandraSerializerCreation)
+  } 
+
+  /**
+   * helper trait for declaring the HList recursive function 
+   * to append keys on an ArrayBuffer in a type safe way
+   */
+  trait Append2Composite[R <: ArrayBuffer[Clause], K <: HList, Q <: HList] {
+    def apply(r: R, k: K, s: List[String], ser: Q): Unit
+  }
+
+  /**
+   * helper implicits for the recursion itself
+   */
+  object Append2Composite {
+    implicit def hnilAppend2Composite[R <: ArrayBuffer[Clause]] = new Append2Composite[R, HNil, HNil] {
+      override def apply(r: R, k: HNil, s: List[String], ser: HNil): Unit = {}
+    }
+    implicit def hlistAppend2Composite[R <: ArrayBuffer[Clause], M, K <: HList, P, Q <: HList](
+      implicit a2c: Append2Composite[R, K, Q]) = new Append2Composite[R, M :: K, P :: Q] {
+    	override def apply(r: R, k: M :: K, s: List[String], ser: P :: Q): Unit = {
+   	      r.add(QueryBuilder.eq(s"""\"${s.head}\"""", ser.head.asInstanceOf[CassandraPrimitive[Any]].toCType(k.head)))
+    	  a2c(r, k.tail, s.tail, ser.tail) 
+    	}
+      }
+  }
+
+  /**
+   * recursive function callee implicit
+   */
+  implicit def append2Composite[R <: ArrayBuffer[Clause], K <: HList, S <: List[String], Q <: HList](r: R)(k: K, s: List[String], ser: Q)
+  	(implicit a2c: Append2Composite[R, K, Q]) = a2c(r, k, s, ser) 
+    
+  /**
+   * helper trait for declaring the HList recursive function 
+   * to create a key from a row
+   * K is the type of HList to return the key, S are the serializers
+   */
+  trait Row2Result[K <: HList, S <: HList] {
```

Comment:
Sorry, I was using `DepFn1` as shorthand for the approach of making the uniquely-determined type parameter a type member. If the `apply` on `DepFn1` doesn't have the right signature, it'd make sense just to change this to `RowResult[K <: HList] { type Out <: HList ...`.